madolson/placeholderkv
# Use io_uring to batch handle clients pending writes to reduce SYSCALL count.

### Description
This patch try to benefit the [io_uring batching feature](https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023#batching) to reduce the SYSCALL count for valkey when `handleClientsWithPendingWrites`. 
With this patch, we can observe more than _**6%**_ perf gain for SET/GET.
This patch was implemented based on below discussion during the review:
1. Introduce a `io_uring.h` to handle the io_uing related API to split it from server logic.
2. Make `io_uring.h` independent of `server.h` .
3. Only use `io_uring` to gain performance when write client static buffer.

## Benchmark Result
### Test Env
- OPERATING SYSTEM: Ubuntu 22.04.4 LTS
- Kernel: 5.15.0-116-generic
- PROCESSOR: Intel Xeon Platinum 8380
- Base: 5b9fc465239036b378a39cbebf3706a2db74dd32
- Server and Client in same socket.

### Test Steps
1. Start valkey-server with below config.
```
taskset -c 0-3 ~/src/valkey-server /tmp/valkey_1.conf

port 9001
bind * -::*
daemonize yes
protected-mode no
save ""
```
5.  Start valkey-benchmark to ensure valkey-server CPU utilized is 1(fully utilized).
```
taskset -c 16-19 ~/src/valkey-benchmark -p 9001 -t set -d 100 -r 1000000 -n 5000000 -c 50 --threads 4
```

### Test Result
QPS of SET and GET can increase  **_6.5%_**, **_6.6%_** correspondingly.

### Perf Stat
The perf stat info shows that only 1 CPU resource was used during the test and the IPC also increase 6%, not from more CPU resources.

```sh
perf stat -p `pidof valkey-server` sleep 10

# w/o io_uring
 Performance counter stats for process id '2267781':

          9,993.95 msec task-clock                #    0.999 CPUs utilized
               625      context-switches          #   62.538 /sec
                 0      cpu-migrations            #    0.000 /sec
            94,933      page-faults               #    9.499 K/sec
    33,894,880,825      cycles                    #    3.392 GHz
    39,284,579,699      instructions              #    1.16  insn per cycle
     7,750,350,988      branches                  #  775.504 M/sec
        73,791,242      branch-misses             #    0.95% of all branches
   169,474,584,465      slots                     #   16.958 G/sec
    39,212,071,735      topdown-retiring          #     23.1% retiring
    11,962,902,869      topdown-bad-spec          #      7.1% bad speculation
    43,199,367,984      topdown-fe-bound          #     25.5% frontend bound
    75,159,711,305      topdown-be-bound          #     44.3% backend bound

      10.001262795 seconds time elapsed

# w/ io_uring
 Performance counter stats for process id '2273716':

          9,970.38 msec task-clock                #    0.997 CPUs utilized
             1,077      context-switches          #  108.020 /sec
                 1      cpu-migrations            #    0.100 /sec
           124,080      page-faults               #   12.445 K/sec
    33,813,062,268      cycles                    #    3.391 GHz
    41,455,816,158      instructions              #    1.23  insn per cycle
     8,063,017,730      branches                  #  808.697 M/sec
        68,008,453      branch-misses             #    0.84% of all branches
   169,066,451,360      slots                     #   16.957 G/sec
    38,077,547,648      topdown-retiring          #     22.0% retiring
    28,509,121,765      topdown-bad-spec          #     16.5% bad speculation
    41,083,738,441      topdown-fe-bound          #     23.8% frontend bound
    65,062,545,805      topdown-be-bound          #     37.7% backend bound

      10.001785198 seconds time elapsed
```

**_NOTE_** 
- Since io_uring was adopted from kernel 5.1, if kernel doesn't support io_uring yet, it will use the origin implementation. 
- This patch introduce the liburing dependency, it is installed in my local env, to keep it simple, I didn't include liburing dependency in this patch, the CI build may failed.
